lisa/tools/lagscope.py

```diff
@@ -0,0 +1,257 @@
+    #      90          0
+    #     105          0
+    #     120          0
+    #     135          1013
+    #     150          7111
+    #     165          15318
+    #     180          35440
+    #     195          39818
+    #     210          63977
+    #     225          51341
+    #     240          62950
+    #     255          78099
+    #     270          88724
+    #     285          76162
+    #     300          52777
+    #     315          76238
+    #     330          93312
+    #     345          141943
+    #     360          62452
+    #     375          32662
+    #     390          10980
+    #     405          3605
+    #     420          1786
+    #     435          992
+    #     450          639
+    #     465          457
+    #     480          2204
+    _interval_frequency_pattern = re.compile(
+        r"\s+(?P<interval_us>\d+)\s+(?P<frequency>\d+)$", re.M
+    )
+
+    @property
+    def dependencies(self) -> List[Type[Tool]]:
+        return [Git]
+
+    @property
+    def command(self) -> str:
+        return "lagscope"
+
+    @property
+    def can_install(self) -> bool:
+        return True
+
+    def _install(self) -> bool:
+        self._install_dep_packages()
+        tool_path = self.get_tool_path()
+        git = self.node.tools[Git]
+        git.clone(self.repo, tool_path, ref=self.branch)
+        code_path = tool_path.joinpath("lagscope")
+        self.node.execute(
+            "./do-cmake.sh build",
+            cwd=code_path,
+            shell=True,
+            expected_exit_code=0,
+            expected_exit_code_failure_message="fail to run do-cmake.sh build",
+        )
+        self.node.execute(
+            "./do-cmake.sh install",
+            cwd=code_path,
+            sudo=True,
+            shell=True,
+            expected_exit_code=0,
+            expected_exit_code_failure_message="fail to run do-cmake.sh install",
+        )
+        self.node.execute(
+            "ln -sf /usr/local/bin/lagscope /usr/bin/lagscope",
+            sudo=True,
+            shell=True,
+            expected_exit_code=0,
+            expected_exit_code_failure_message="fail to create symlink to lagscope",
+        )
+        return self._check_exists()
+
+    def run_as_server(self, ip: str = "", daemon: bool = True) -> None:
+        # -r: run as a receiver
+        # -rip: run as server mode with specified ip address
+        # -D: run as a daemon
+        cmd = ""
+        if daemon:
+            cmd += " -D"
+        if ip:
+            cmd += f" -r{ip}"
+        else:
+            cmd += " -r"
+        self.run(
+            cmd,
+            force_run=True,
+            sudo=True,
+            shell=True,
+            expected_exit_code=0,
+            expected_exit_code_failure_message=f"fail to launch cmd {self.command}"
+            f"{cmd}",
+        )
+
+    def run_as_client(
+        self,
+        server_ip: str,
+        test_interval: int = 0,
+        ping_count: int = 1000000,
```

Comment:
The default value is too big to me. Does it mean to ping 1 million times?

Reply:
not sure, I know it will cost 5 mins to finish testing,

Reply:
please double check it, and also find a smaller number.

Reply:
yes, it means ping 1 million times

Reply:
We can set a default smaller value, but for perf testing, I still insist on using this.

Reply:
what's value of a big number? The different perf test case has different goal. And we need to get result faster and cheaper. For latency testing, it just needs the average data. For stress or pressure testing, we can run high concurrency. For long haul testing, we run it long time. But the normal perf test cases will be run in many places, we need minimize test scope. So the default value should be minimized, and other stress/pressure tests can overwrite it.

Reply:
it is quite quick (1 second) but with below error, I need update to a proper value.
```02:42:54 INFO: failed to bind socket[3] to a local port: [0.0.0.0:25001]. errno = 98. Ignored```

Reply:
the default value is 1000000, or I can add test duration, to run 2 mins.

![image](https://user-images.githubusercontent.com/10083705/149442867-10100ca1-59ec-4142-bb97-de38338064a3.png)

Reply:
we can try a test duration like 10 seconds. The ping iteration may run too fast to get errors.

Reply:
update to 10 seconds (around 45000 times ping)